cdap-master/src/main/java/co/cask/cdap/data/tools/ReplicationStatusTool.java

```diff
@@ -0,0 +1,379 @@
+        ByteBuffer row = ByteBuffer.wrap(result.getRow());
+        getNextPart(row); // Ignore cluster ID;
+        String region = new String(getNextPart(row), "UTF-8");
+        long timestamp = getTimeFromResult(result, qualifier);
+
+        if (timeMap.get(region) == null || timestamp > timeMap.get(region)) {
+          timeMap.put(region, timestamp);
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Error while reading table:" + e);
+      throw Throwables.propagate(e);
+    } finally {
+      hTable.close();
+    }
+    return timeMap;
+  }
+
+  private static byte[] getNextPart(ByteBuffer byteBuffer) {
+    int len = byteBuffer.getInt();
+    if (byteBuffer.remaining() < len) {
+      throw new BufferUnderflowException();
+    }
+    byte[] bytes = new byte[len];
+    byteBuffer.get(bytes, 0, len);
+    return bytes;
+  }
+
+  private static TableId getReplicationStateTableId(HBaseTableUtil tableUtil) throws IOException {
+    String tableName = hConf.get(ReplicationConstants.ReplicationStatusTool.REPLICATION_STATE_TABLE_NAME);
+    String ns = hConf.get(ReplicationConstants.ReplicationStatusTool.REPLICATION_STATE_TABLE_NAMESPACE);
+    TableId tableId = tableUtil.createHTableId(
+      (ns != null)
+        ? new NamespaceId(ns)
+        : new NamespaceId(ReplicationConstants.ReplicationStatusTool.REPLICATION_STATE_TABLE_DEFAULT_NAMESPACE),
+      (tableName != null)
+        ? tableName
+        : ReplicationConstants.ReplicationStatusTool.REPLICATION_STATE_TABLE_DEFAULT_NAME);
+    return tableId;
+  }
+
+  private static long getTimeFromResult(Result result, String qualifier) {
+    long timestamp = 0L;
+    List<Cell> timeVal = result.getColumnCells(Bytes.toBytes(ReplicationConstants.ReplicationStatusTool.TIME),
+                                               Bytes.toBytes(qualifier));
+    if (timeVal.size() > 0) {
+      ByteBuffer timeBuffer =  ByteBuffer.wrap(CellUtil.cloneValue(timeVal.get(0)));
+      timestamp = timeBuffer.getLong();
+    }
+    return timestamp;
+  }
+
+  private static Long getShutdownTime() throws IOException {
+    Long shutdownTime = 0L;
+
+    try (BufferedReader br = new BufferedReader(new FileReader(Constants.Replication.CDAP_SHUTDOWN_TIME_FILENAME))) {
+      //Read the shutdown Time
+      shutdownTime = new Long(br.readLine());
+    } catch (FileNotFoundException e) {
+      LOG.error("Cannot read shutdown time.");
+    }
+    LOG.info("Read CDAP master shutdown time:" + shutdownTime);
+    return shutdownTime;
+  }
+
+  /**
+   * This function implements the logic to test if replication is complete for a given table
+   * @param shutdownTime CDAP shutdown time
+   * @param writeTime timestamp of the last WAL entry written for this table
+   * @param replicateTime timestamp of the last WAL entry replicated for this table
+   * @return
+   */
+  private static boolean isReplicationComplete(long shutdownTime, long writeTime, long replicateTime) {
```

Comment:
I was trying to keep the main deciding logic separate also leaving room for any changes later on.